pkg/kv/kvserver/logstore/sideload_disk.go

```diff
@@ -276,3 +293,57 @@ func (ss *DiskSideloadStorage) String() string {
 	fmt.Fprintf(&buf, "(%d files)\n", count)
 	return buf.String()
 }
+
+// mkdirAllAndSyncParents creates the given chain of directories, if any
+// subdirectory is missing, and syncs parents of all the created directories.
+// The base directory must already exist. The directories are created using the
+// provided permissions mask, with the same semantics as in os.MkdirAll.
+//
+// For example, if dirs is [x, y, z], then this func creates all the missing
+// subdirectories in base/x/y/z. If base is the lowest previously existing
+// directory then base, x, and y are synced. If x existed too, then only x and y
+// are synced. If y existed before then there are no syncs.
+//
+// NB: path syntax in the examples above may differ, because components of the
+// path are combined using the provided filesystem interface.
+func mkdirAllAndSyncParents(fs vfs.FS, base string, dirs []string, perm os.FileMode) error {
```

Comment:
This is a modified version of a similar [func](https://github.com/cockroachdb/pebble/blob/124486b0f85d75a0c695cb65703c5495eec66c99/checkpoint.go#L87) in pebble.

I trust this implementation more though, because:
- there is no sketchy "." checks (which, if done wrong, can lead to infinite loops)
- there is a unit-test

@cockroachdb/storage consider upstreaming this and/or the tests.

Reply:
Any compelling reason why this doesn't simply take the path as a string, without special-casing `base`, and traverse it using e.g. `os.filepath.Dir()` until it finds an existing directory?

I see you mention sketchy `.` checks, but what's preventing the caller from passing in a path with `.` here anyway? Seems better to just handle that case.

Reply:
The other implementation does:
```
while path != "." {
	process(path)
	path = parent(path) // who guarantees that this will always return "." in the end?
}
```

The current implementation always terminates because the path is already decomposed and has a predefined number of elements.